operators/tokenizer/bpe_utils.hpp

```diff
@@ -114,12 +116,115 @@ class TokenWithRegularExp {
 
-    // 's|'t|'re|'ve|'m|'ll|'d|
+  /*
+  std::regex does not directly support std::u32string (which is what U"" literals produce).
+  The std::regex class template is specialized for char and wchar_t types, but not for char32_t or char16_t.
+  To work with Unicode strings, we therefore convert the std::u32string to a std::wstring.
+
+  Wide strings, or std::wstring, are used for supporting a large character set, including Unicode
+  characters and characters from various languages.
+  */
+
+  static std::wstring U2Wstring(const std::u32string& u32str) {
+    std::wstring wstr;
+    wstr.reserve(u32str.size()); // Reserve space to avoid multiple allocations
+
+    for (char32_t codepoint : u32str) {
+      if (codepoint <= 0xFFFF) {
+        // Single UTF-16 code unit
+        wstr.push_back(static_cast<wchar_t>(codepoint));
+      } else if (codepoint <= 0x10FFFF) {
+        // Surrogate pair
+        codepoint -= 0x10000;
+        wchar_t high_surrogate = static_cast<wchar_t>((codepoint >> 10) + 0xD800);
+        wchar_t low_surrogate = static_cast<wchar_t>((codepoint & 0x3FF) + 0xDC00);
+        wstr.push_back(high_surrogate);
+        wstr.push_back(low_surrogate);
+      } else {
+        // Invalid code point for UTF-16
+        throw std::runtime_error("Invalid UTF-32 codepoint encountered");
+      }
+    }
+
+    return wstr;
+  }
+
+
+  std::u32string W2Ustring(const std::wstring& wstr) {
+      std::u32string u32str;
+      u32str.reserve(wstr.size());  // Reserve space to avoid multiple allocations
+
+      for (wchar_t wc : wstr) {
+          if (wc <= 0x7F) {
+              // 1-byte character (ASCII)
+              u32str.push_back(static_cast<char32_t>(wc));
+          } else if (wc <= 0x7FF) {
+              // 2-byte character
+              char32_t ch = (static_cast<char32_t>(wc) & 0x07FF) | 0x0800;
+              u32str.push_back(ch);
+          } else if (wc <= 0xFFFF) {
+              // 3-byte character
+              char32_t ch = (static_cast<char32_t>(wc) & 0x0FFF) | 0xD800;
+              u32str.push_back(ch);
+              ch = (static_cast<char32_t>(wc) >> 10) | 0xDC00;
+              u32str.push_back(ch);
+          } else if (wc <= 0x10FFFF) {
+              // 4-byte character (surrogate pairs)
+              char32_t ch = ((wc >> 10) & 0x3FF) | 0xD800;
+              u32str.push_back(ch);
+              ch = (wc & 0x3FF) | 0xDC00;
+              u32str.push_back(ch);
+          } else {
+              // Invalid Unicode code point
+              throw std::runtime_error("Invalid wide character encountered");
+          }
+      }
+
+      return u32str;
+  }
+
+  // Use the C++ standard library to handle regex pattern matching for compatible strings
+  std::u32string_view RegexMatchSTD(const std::u32string& regex) {
+    static std::u32string text(m_text);
+    std::wstring wstr = U2Wstring(text);
+    std::wstring wpattern = U2Wstring(regex);
+
+    std::wregex pattern(wpattern);
```

Comment:
why do we have to use the wchar edition of std::regex? Since it doesn't handle unicode code, std::regex should be fine too, right?

Reply:
even if the pattern is not in unicode, if the text to pattern match with is in unicode, we need to use wstrings, hence we use wregex and have to convert to and fro, because although the pattern does not have unicode chars, the input text does.

Reply:
That's actually an interested topic. If a pattern is utf-8 encoded string, and it matches well on utf-8 input text as well if you take some time to think about it.

Reply:
Yep! I already had some ideas for a related PR, can discuss with you in our weekly sync.